File: leetcode/c++/LCR_194_er_cha_shu_de_zui_jin_gong_gong_zu_xian_lcof.cpp
// https://leetcode.cn/problems/er-cha-shu-de-zui-jin-gong-gong-zu-xian-lcof/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == NULL || root == p || root == q)
            return root;
        TreeNode* l = lowestCommonAncestor(root->left, p, q);
        TreeNode* r = lowestCommonAncestor(root->right, p, q);
        if (l == NULL && r == NULL) 
            return NULL;
        else if (l == NULL)
            return r;
        else if (r == NULL)
            return l;
        else 
            return root;
    }
};

// O(n), O(n)

File: leetcode/golang/LCR_194_er_cha_shu_de_zui_jin_gong_gong_zu_xian_lcof.go
type TreeNode struct {
         Val int
         Left *TreeNode
         Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    if root == nil || root == p || root == q {
        return root
    }
    l, r := lowestCommonAncestor(root.Left, p, q), lowestCommonAncestor(root.Right, p, q)
    if l == nil && r == nil {
        return nil
    } else if l == nil {
        return r
    } else if r == nil {
        return l
    } else {
        return root
    }
}